### PR TITLE
UI: fix #1153 as well as #1185

### LINKED ISSUE #1153: c-bench: `rel err` column sort order broken when "n/a" values are present
### unsorted:
image

### sorted:
image

8.2 is not the largest!

### PR DESCRIPTION
**Pull request by jgehrcke**
Fixes #1153 as well as #1185.

Reference docs: https://datatables.net/examples/advanced_init/html5-data-attributes.html

See title.

---

### desc
Screenshot from 2023-04-27 13-17-05

### asc

Screenshot from 2023-04-27 13-17-01

### CODE REVIEW

File: conbench/templates/c-benchmark-results-for-case.html [hunk 1]
```diff
@@ -44,7 +44,7 @@ <h3>
               <td class="font-monospace">{{ result.context.id[:8] }}</td>
               <td class="font-monospace">{{ result.ui_non_null_sample_count }}</td>
               <td class="font-monospace">{{ result.ui_mean_and_uncertainty }}</td>
-              <td class="font-monospace">{{ result.ui_rel_sem }}</td>
+              <td class="font-monospace" data-order="{{ result.ui_rel_sem[0] }}">{{ result.ui_rel_sem[1] }}</td>
```
Comment (jgehrcke) [inline conbench/templates/c-benchmark-results-for-case.html:47]: TODO: make this not be two function calls

**Comment by jgehrcke** [comment]
:ship:

File: conbench/app/_plots.py [hunk 2]
```diff
@@ -571,11 +580,14 @@ def time_series_plot(
...
     )
 
-    source_current_bm_mean = _source(
-        [dummy_hs_for_current_svs],
-        unit,
-        formatted=formatted,
-    )
+    # Edge case: this may be failed, in which case we cannot show a data point!
+    source_current_bm_mean = "failed"
+    if not current_benchmark_result.is_failed():
```
Comment (jgehrcke) [inline conbench/app/_plots.py:585]: Previously, we showed this data point, I think, despite the benchmark result being "failed".

Interesting, CC @jonkeane

File: conbench/app/_plots.py [hunk 3]
```diff
@@ -78,13 +78,22 @@ def get_history_plot(
         """
         samples = get_history_for_benchmark(benchmark_result_id=benchmark_result.id)
 
+        # Does `samples` include the current benchmark result? Interesting: the
```
Comment (jgehrcke) [inline conbench/app/_plots.py:81]: answer is "no" I think, but should probably be spelled out explicitly in the docstring(s) of the history-related function(s).

File: conbench/app/_plots.py [hunk 4]
```diff
@@ -78,13 +78,22 @@ def get_history_plot(
...
+        # current benchmark result may be failed, but I think all items in
+        # samples (historical benchmark results are guaranteed to not be
+        # failed.
```
Comment (jgehrcke) [inline conbench/app/_plots.py:84]: Interesting consideration, CC @austin3dickey -- we should also specify this properly in related docstring(s).

File: conbench/entities/benchmark_result.py [hunk 5]
```diff
@@ -486,23 +486,30 @@ def ui_mean_and_uncertainty(self) -> str:
 
     # maybe make this a cached property
     @property
-    def ui_rel_sem(self) -> str:
+    def ui_rel_sem(self) -> Tuple[str, str]:
+        """
+        The first string is a stringified float for sorting in the table.
+        The second string is for display in the table, with unit.
+        Associate "n/a" with a negative value so that DESC sorting by error
+        is helpful.
```
Comment (jgehrcke) [inline conbench/entities/benchmark_result.py:494]: Will need to make this interface better in the (near) future. This works for now, moving on.

File: conbench/app/_plots.py (hunk 4 above)
Comment (austin3dickey) [reply conbench/app/_plots.py:84]: Yeah! We do specify that history samples always consist of non-errored benchmark results:
https://github.com/conbench/conbench/blob/6b12e2b7152840105c379341ed48a94ba451c4ad/conbench/entities/history.py#L188
but I don't think we wrote down that you can still get history _given_ an errored benchmark result

**Pull request merged**